### PR TITLE
[processor/attributes] Support attributes set by server authenticators

### LINKED ISSUE #9241: Attributes processor from_context should provide access to AuthData attributes
**Is your feature request related to a problem? Please describe.**
`ServerAuthenticator` extensions such as BasicAuth and OIDC store `client.AuthData` interface implementations containing attributes describing the authenticated client as `client.Info.Auth`. Attributes processor actions using `from_context` can create span attributes from `client.Info.Metadata`, but not from `client.Info.Auth`. There are use cases where adding `AuthData` attributes such as the authenticated username to spans is needed, but this is not currently possible.

**Describe the solution you'd like**
The Attributes processor should be extended to also extract `from_context` attributes from `client.Info.Auth`. Either `client.Info.Auth` would be searched when the key is not found in `client.Info.Metadata`, or keys beginning with a prefix such as `"Auth."` would be searched for only in `client.Info.Auth`.

**Describe alternatives you've considered**
No existing processor provides the ability to add span attributes based on data from `client.Info.Auth`. A new processor could be added for this, but it would largely duplicate the functionality of the existing Attributes processor.

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by svrakitin**
**Description:**

Support looking up attributes from `client.Info.Auth` which is set by extensions implementing `configauth.ServerAuthenticator`. This way we can attach authentication of the exporter (e.g. OIDC subject) to the telemetry.

User can refer to these attributes by prefixing keys with `auth.` when using `from_context`:

```yaml
attributes/subject:
  actions:
  - key: enduser.id
    from_context: auth.subject
    action: insert
```

For consistency I added `metadata.` prefix to refer to actual metadata/header keys. This prefix is optional for backwards compatibility.

**Testing:**

I added tests to `attraction` package which has actual extraction logic.

Closes #9241

### CODE REVIEW

File: internal/coreinternal/attraction/attraction.go
Comment (svrakitin) [inline internal/coreinternal/attraction/attraction.go:340]: What to do with non-string typed attributes? At the moment we don't have any use cases as auth extensions either return `string` or `[]string` attributes.

Should I match on raw type and produce corresponding value? I would probably do that if `newValueFromRaw` was public. https://github.com/open-telemetry/opentelemetry-collector/blob/6fed15bf12d807da8ac834fd776eb1423187ca19/pdata/internal/common.go#L135

Comment (pmm-sumo) [reply internal/coreinternal/attraction/attraction.go:340]: I wonder if this might ever happen. Perhaps we should just log that unexpected type was observed (ideally only once for a while, e.g. using logging sampler or other means)

Comment (jpkrohling) [reply internal/coreinternal/attraction/attraction.go:340]: I consider this an error condition and users should be appropriately warned if this ever happens.

File: processor/attributesprocessor/README.md
Comment (jpkrohling) [inline processor/attributesprocessor/README.md:49]:
```suggestion
  # If the key is prefixed with `metadata.`, the values are searched 
```

If we ever intend to remove the support for unqualified keys, it's better to not mention the "no prefix" here.

File: processor/attributesprocessor/README.md
Comment (jpkrohling) [inline processor/attributesprocessor/README.md:50]:
```suggestion
  # in the receiver's transport protocol additional information like gRPC Metadata or HTTP Headers. 
```

File: processor/attributesprocessor/README.md
Comment (jpkrohling) [inline processor/attributesprocessor/README.md:51]:
```suggestion
  # If the key is prefixed with `auth.`, the values are searched
```

File: processor/attributesprocessor/README.md
Comment (jpkrohling) [inline processor/attributesprocessor/README.md:52]:
```suggestion
  # in the authentication information set by the server authenticator. Refer to the server authenticator's documentation part of your pipeline for more information about which attributes are available.
```

File: processor/attributesprocessor/testdata/config.yaml
Comment (jpkrohling) [inline processor/attributesprocessor/testdata/config.yaml:351]: Can you refer to a specific server authenticator that produces `subject`?

**Review by jpkrohling — COMMENTED** [review]
LGTM, just the documentation could be a bit better. Thanks for this PR!

**Comment by svrakitin** [comment]
@jpkrohling I added logger to the `AttrProc` and removed it from the `Process` method. Now logger is injected from the processor.

Thanks for suggestions to the docs.

File: processor/attributesprocessor/testdata/config.yaml
Comment (svrakitin) [reply processor/attributesprocessor/testdata/config.yaml:356]: I hope this is useful enough. Probably needs help with wording.

File: internal/coreinternal/attraction/attraction.go
Comment (svrakitin) [reply internal/coreinternal/attraction/attraction.go:353]: Added log warning.

File: internal/coreinternal/attraction/attraction.go
Comment (pmm-sumo) [inline internal/coreinternal/attraction/attraction.go:347]: This could potentially generate a lot of log lines. We've previously recommended using (to the extent of copy/pasting) [sampled logger](https://github.com/open-telemetry/opentelemetry-collector/blob/main/exporter/exporterhelper/queued_retry.go#L61-L78) though with so many cases it's relevant perhaps we just put it into some common helper package

Comment (svrakitin) [reply internal/coreinternal/attraction/attraction.go:347]: Thanks for the tip, I missed that previously.

File: internal/coreinternal/attraction/attraction_test.go
Comment (jpkrohling) [inline internal/coreinternal/attraction/attraction_test.go:89]: I won't block the PR for this unless @pmm-sumo agrees with me, but I would prefer the test to be passing zap.NewNop() instead of nil. Remember that your test code shows the good practice for consumers of the main code, and by passing nil, you are sending the message that nil is OK, when it really isn't.

Comment (svrakitin) [reply internal/coreinternal/attraction/attraction_test.go:89]: Sure, I don't have a preference here. Gonna update.

File: internal/coreinternal/attraction/attraction.go
Comment (pmm-sumo) [inline internal/coreinternal/attraction/attraction.go:180]: I think now we might assume that logger is never nil?

**Review by pmm-sumo — DISMISSED** [review]

**Review by rkukura — APPROVED** [review]
This all looks good to me - very similar to what I planned to implement when filing https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/9241, except that I didn't anticipate the logging changes. Please let me know if I can help get this merged.

File: internal/coreinternal/attraction/attraction.go
Comment (jpkrohling) [reply internal/coreinternal/attraction/attraction.go:180]: It's good to check for nil, but not sure about the outcome: you do want to have access to a logger, it's not an optional parameter, so, you should return an error if it's nil.

File: internal/coreinternal/attraction/attraction.go
Comment (jpkrohling) [inline internal/coreinternal/attraction/attraction.go:372]: What's this line doing?

**Review by jpkrohling — APPROVED** [review]

**Comment by jpkrohling** [comment]
PR rebased

**Comment by svrakitin** [comment]
@jpkrohling What is left to get it merged?

**Comment by pmm-sumo** [comment]
Juraci is now on parental leave :) But since we have 3 approvals (including two from approvers/contributors) and passing tests, I'm marking this as `ready-to-merge`

File: internal/coreinternal/attraction/attraction.go
Comment (bogdandrutu) [inline internal/coreinternal/attraction/attraction.go:285]: Why this random configuration?

File: internal/coreinternal/attraction/attraction.go
Comment (bogdandrutu) [reply internal/coreinternal/attraction/attraction.go:180]: I agree, this seems to be unrelated, consider to split a separate PR.

File: internal/coreinternal/attraction/attraction.go
Comment (bogdandrutu) [inline internal/coreinternal/attraction/attraction.go:311]: This was fixed in a separate PR, and also unrelated to your main PR.

**Comment by bogdandrutu** [comment]
@pmm-sumo there are lots of unrelated changes, and some random configurations for logger (why do we need sampling here and not in other places? add debug logs if too many, record metrics, etc.)

**Comment by svrakitin** [comment]
@bogdandrutu There was a suggestion to log invalid atrribute types if they are returned by ServerAuthenticator: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/9420#discussion_r858965522

I don’t care much about sampling and can drop it. It was taken from the suggestion in one of the comments above.

I will reduce this PR to changes I originally introduced and leave further changes to maintainers who have a better idea than me.

**Comment by bogdandrutu** [comment]
@svrakitin it is not about "better idea than me", it is about keeping the PRs focus on what we want to implement :)

**Comment by pmm-sumo** [comment]
Good point @bogdandrutu, apologies for messing up with tagging and thanks for bringing this up. 

On the sampled logger thing, I believe it [stems from this comment](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/9420/files#r859067416). I mentioned sampled logging since this was something brought to my attention in the past (don't recall which PR but I can look it up for context).

In principle, I think that user should know that something is wrong and because of that `debug` logs are not a good fit.  I also heard some comments that the collector is sometimes too quiet over what happens in the components and some users anticipate more logging. At the same time, we want to limit the impact of that log if happens frequently, hence (aggressive) sampled logger and `warn` level. 

So we have two paths here:
* always use `debug` level for such kind of logs
* use sampled logger and move it to a common package, so other components could use it easily for potentially noisy logs

I was inclined towards the latter (in a separate PR). There are some components doing that already and I was aiming at doing the cleanup after this get merged. Before I open the issue wanted to check what is your preference @bogdandrutu

**Review by bogdandrutu — COMMENTED** [review]
This PR looks much better now, @pmm-sumo please re-approve since you are familiar, and add the label again

**Review by pmm-sumo — APPROVED** [review]

**Pull request merged**
